### PR TITLE
FW navigation first order hold move to position controller

### PR DESCRIPTION
**Pull request by dagar**
Take 2 https://github.com/PX4/Firmware/pull/8883

### CODE REVIEW

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 1]
```diff
@@ -694,7 +773,7 @@ FixedwingPositionControl::control_position(const hrt_abstime &now, const Vector2
 						   false,
 						   radians(_param_fw_p_lim_min.get()));
 
-		} else if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
+		} else if (position_sp_type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
```
Comment (RomanBapst) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:776]: @dagar This is a bit confusing. Above you are using `pos_sp_curr.type` while here you are using `position_sp_type`

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -673,18 +673,97 @@ FixedwingPositionControl::control_position(const hrt_abstime &now, const Vector2
...
+					position_sp_type = position_setpoint_s::SETPOINT_TYPE_LOITER;
+				}
+
+			} else if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
+				// LOITER: use SETPOINT_TYPE_POSITION to get to SETPOINT_TYPE_LOITER
+				if ((dist >= 0.f)
+				    && (dist_z > 2.f * _param_fw_clmbout_diff.get())
+				    && (dist_xy > 2.f * math::max(acc_rad, fabsf(pos_sp_curr.loiter_radius)))) {
+					// SETPOINT_TYPE_LOITER -> SETPOINT_TYPE_POSITION
+					position_sp_type = position_setpoint_s::SETPOINT_TYPE_LOITER;
```
Comment (RomanBapst) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:714]:
```suggestion
					position_sp_type = position_setpoint_s::SETPOINT_TYPE_POSITION;
```

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -673,18 +673,97 @@ FixedwingPositionControl::control_position(const hrt_abstime &now, const Vector2
...
+				}
+			}
+		}
+
+
 		if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_IDLE) {
```
Comment (RomanBapst) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:720]:
```suggestion
		if (position_sp_type == position_setpoint_s::SETPOINT_TYPE_IDLE) {
```

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -673,18 +673,97 @@ FixedwingPositionControl::control_position(const hrt_abstime &now, const Vector2
...
 			_att_sp.thrust_body[0] = 0.0f;
 			_att_sp.roll_body = 0.0f;
 			_att_sp.pitch_body = 0.0f;
 
 		} else if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION) {
```
Comment (RomanBapst) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:725]:
```suggestion
		} else if (position_sp_type == position_setpoint_s::SETPOINT_TYPE_POSITION) {
```

**Comment by RomanBapst** [comment]
@dagar I did some testing on this branch. I fixed a few things that seemed obviously wrong to me (see comments above). There's still an issue in the logic switching between position setpoint and loiter setpoint. It seems to constantly switch between the two. I need to dig a bit deeper.

**Comment by RomanBapst** [comment]
@dagar Please check the commit I pushed.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:776]: No real reason, nasty rebase.

**Comment by RomanBapst** [comment]
@dagar I fixed a bug in the FOH logic and did some small changes to make things work. I don't really like this too much though. The main problem is that we decide to switch from a position setpoint to a loiter based on an acceptance radius. However, the loiter radius might be much bigger than the acceptance radius and so as the vehicle is trying to loiter the logic will put it back into position mode.
Currently I solved the problem by comparing acceptance radius with L1 distance and using the max of the two but that means that the FOH logic will also not work exactly as expected in some cases.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -718,6 +720,8 @@ FixedwingPositionControl::control_position(const hrt_abstime &now, const Vector2
 			}
 		}
 
+		_type = position_sp_type;
```
Comment (dagar) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:723]: This should probably probably be reset for the other cases.

**Comment by dagar** [comment]
Thanks @RomanBapst, looks good.

**Pull request merged**